docs/rfcs/20220413-aot-for-all-snode.md

```diff
@@ -0,0 +1,340 @@
+
+```py
+x = ti.Vector.field(3, ti.f32)
+for f in x._get_field_members():  # `x` consists three scalar f32 fields
+  ti.root.dense(ti.ij).place(f)
+```
+
+This introduces confusion at several spots:
+
+1. Type is not purely decided by `dtype`, but also by how the field is placed.
+2. It introduces the notion of "nested field", which Taichi doesn't currently have a good abstraction for. Because of this, it is quite complicated to apply certain kind of optimizations for a composite-typed field. For example, vectorized load/save consumes the same bandwidth as scalar ops on certain platforms. Without a good abstraction, the checking for whether a matrix field is AoS or SoA has to be spread across different passes in CHI IR.
+
+If we further think about the problem, SoA `x` is not really a field. Instead, it is a *grouped view* of three individual scalar fields. This view provides matrix operations, which won't make sense for each individual scalar field.
+
+In addition, because type is currently coupled with Taichi field definition, a Taichi field has to be implemented as individual fields in order to support the SoA scenario. Once we switch to the type builder pattern, we can control how the type is constructed first, and choose the field implementation later.
+
+
+```py
+builder = ti.SNodeTreeBuilder()
+vel = builder.add_field(dtype=ti.vec3, name='vel', layout=ti.Layout.SOA)
+
+for c in vel.compoenents():
+  builder.tree().dense(ti.ij, 4).place(c)
+```
+
+If we want to make it explicit that this is a *field view*, we can do the following:
+
+```py
+builder = ti.SNodeTreeBuilder()
+v0 = builder.add_field(dtype=ti.f32)
+v1 = builder.add_field(dtype=ti.f32)
+v2 = builder.add_field(dtype=ti.f32)
+for v in [v0, v1, v2]:
+  builder.tree().dense(ti.ij, 4).place(v)
+
+# Checks that
+# 1. `components` and `dtype` are compatible.
+# 2. All the fields in `components` are homogeneous in their SNode hierarchy.
+# 3. `components` themselves are not field views.
+vel = builder.add_field_view(dtype=ti.vec3, name='vel', components=[v0, v1, v2])
+```
+
+Matrix field view supports common matrix operations, and is equivalent to expanding each component into a local matrix variable.
+
+```py
+# 1
+vel_soa[i, j].inverse()
+# equivalent to
+ti.vec3([v0[i, j], v1[i, j], v2[i, j]]).inverse()
+
+# 2
+vel_soa[i, j][1] += 2.0
+# equivalent to
+v1[i, j] += 2.0
+
+# 3
+vel_soa[i, j] = vel_soa[i, j] @ some_vec3
+# equivalent to
+vel_tmp = ti.vec3([v0[i, j], v1[i, j], v2[i, j]])
+vel_tmp = vel_tmp @ some_vec3
+v0[i, j] = vel_tmp[0]
+v1[i, j] = vel_tmp[1]
+v2[i, j] = vel_tmp[2]
+```
+
+## Python AOT API
+
+Here's the Python AOT API to save the SNodeTree type.
+
+```py
+builder = ti.SNodeTreeBuilder()
+# ...
+tree_t = builder.build()
+
+@ti.kernel
+def foo(tr: tree_t):
+  # ...
+
+m = ti.aot.Module(arch)
+m.add_snode_tree_type(tree_t, name="vel_tree")
+m.add_kernel(foo)
```

Comment:
As we discussed offline this RFC is mainly about how to handle the global fields from our python program in aot. As a result the ongoing effort for compute graph will mainly focus on replacing `add_kernel` as `add_graph` (or a similar name) and not worry about fields for now.